_posts/2019-07-25-testy-frontendu-okiem-full-stacka.md

```diff
@@ -0,0 +1,136 @@
+Mając przykładowy komponent:
+```typescript
+@Component({
+    selector: 'date-picker',
+    template: `
+        <date-picker-input [value]="formattedValue" (change)="onValueChange($event)"></date-picker-input>
+    `
+})
+export class DatePicker implements OnChanges  {
+    @Input value: string;
+    formattedValue: string;
+
+    constructor(private service: FormatterService ) {
+    }
+    
+    ngOnChanges(changes: SimpleChanges): void {
+        this.formattedValue = this.service.format(changes.value.currentValue);
+    }
+
+    onValueChange(event: string): void {
+        // ...
+    }
+}
+```
+
+test serwisu formattera mógłby wyglądać następująco:
+```typescript
+describe('FormatterService', () => {
+
+   beforeEach(() => {
+      service = new FormatterService();
+   });
+
+   it('should return formatted date from timestamp', () => {
+      // given
+      const timestamp: string = Date.now('2018-07-01').toString(); 
+
+      // when
+      const result = service.format(timestamp);
+
+      // then
+      expect(result).toBe('1 lipca 2018');
+   });
+
+});
+```
+
+Zaprojektowany w ten sposób komponent pozwolił na przetestowanie głównej funkcjonalności nie przejmując się zależnościami komponentu i jego szablonem. Tak wyglądałyby testy, jeśli zaniedbalibyśmy wyżej zaproponowany podział:
+komponent: 
+```typescript
+@Component({
+    selector: 'date-picker',
+    template: `
+        <input [value]="formattedValue" (change)="onValueChange($event.target.value)">
+    `
+})
+export class DatePicker implements OnChanges  {
+    @Input value: string;
+    formattedValue: string;
+
+
+    constructor(private service: FormatterService ) {
+    }
+    
+    ngOnChanges(changes: SimpleChanges): void {
+        this.formattedValue = this.format(changes.value.currentValue);
+    }
+ 
+    onValueChange(event: string): void {
+        ...
+    }
+
+
+    private format(string: value): string {
+        ...
+    }
+}
+```
+
+test:
+```typescript
+let fixture: ComponentFixture<DatePicker>;
+describe('DatePicker', () => {
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            declarations: [DatePicker]
+        });
+        fixture = TestBed.createComponent(DatePicker);
+    });
+
+    it('should return formatted date from timestamp', () => {
+        // given
+        const hostElement = fixture.nativeElement;
+        const input: HTMLInputElement = hostElement.querySelector('input');
+
+        // when
+        fixture.componentInstance.value = '2018-07-01';
+        fixture.detectChanges();
+
+        // then
+        expect(input.value).toBe('1 lipca 2018');
+    });
+});
+```
+Jak widać testy znacząco spuchły, utraciły na czytelności oraz stały się zależne od szablonu. Różnica w liczbie linijek może nie jest kolosalna, ale wraz z dodawaniem funkcjonalności i zależności będzie coraz trudniej będzie utrzymać klarowność. Jednocześnie kuszące jest, aby upublicznić metodę "format" tylko na potrzebny testów.
```

Comment:
Trochę mi się nie podoba struktura artykułu. Najpierw piszemy, że warto pisać testy i trudno znaleźć wymówki żeby tego nie robić, następnie pokazujemy co w jaki sposób testować i mamy przykład słabej struktury kodu potem zrefaktorowanej do lepszej z punktu widzenia testów.

1. Brakuje mi miejsca, w którym powiemy "Patrzcie! to jest takie proste!"
2. Tak nie do końca jestem przekonany co do przykładu, bo temat dekompozycji kodu został trochę spłycony i osoba czytająca ten tekst może zrozumieć, że "warto dekomponować kod, bo wtedy się go łatwiej testuje"
3. Przydałoby się jakieś zdanie podsumowania

Reply:
1. znajdę odpowiednie miejsce i dorzucę 2-3 zdania
2. podziałam w tym temacie, znajdę bardziej przekonujący przykład, który nie będzie sugerował dekompozycji
3. dorzucę coś